File: app-java/src/test/java/denshchikov/dmitry/taskcreator/config/IntegrationTest.java
package denshchikov.dmitry.taskcreator.config;

import org.springframework.test.context.DynamicPropertyRegistry;
import org.springframework.test.context.DynamicPropertySource;
import org.testcontainers.containers.PostgreSQLContainer;
import org.testcontainers.containers.wait.strategy.Wait;
import org.testcontainers.containers.wait.strategy.WaitAllStrategy;
import org.testcontainers.junit.jupiter.Container;
import org.testcontainers.junit.jupiter.Testcontainers;

import java.time.Duration;

import static org.testcontainers.containers.wait.strategy.WaitAllStrategy.Mode.WITH_MAXIMUM_OUTER_TIMEOUT;

@Testcontainers
public interface IntegrationTest {

  @Container
  PostgreSQLContainer<?> postgresContainer =
      new PostgreSQLContainer<>("postgres:latest")
          .waitingFor(
              new WaitAllStrategy(WITH_MAXIMUM_OUTER_TIMEOUT)
                  .withStartupTimeout(Duration.ofSeconds(90))
                  .withStrategy(
                      Wait.forLogMessage(
                          ".*database system is ready to accept connections.*\\s", 2))
                  .withStrategy(Wait.forListeningPort()));

  @DynamicPropertySource
  static void datasourceProperties(DynamicPropertyRegistry registry) {
    registry.add("spring.datasource.username", postgresContainer::getUsername);
    registry.add("spring.datasource.password", postgresContainer::getPassword);
    registry.add("spring.datasource.url", postgresContainer::getJdbcUrl);
  }
}

File: src/test/kotlin/denshchikov/dmitry/taskcreator/config/IntegrationTest.kt
package denshchikov.dmitry.taskcreator.config

import org.junit.jupiter.api.BeforeAll
import org.springframework.test.context.DynamicPropertyRegistry
import org.springframework.test.context.DynamicPropertySource
import org.testcontainers.containers.PostgreSQLContainer
import org.testcontainers.containers.wait.strategy.Wait
import org.testcontainers.containers.wait.strategy.WaitAllStrategy
import org.testcontainers.containers.wait.strategy.WaitAllStrategy.Mode.WITH_MAXIMUM_OUTER_TIMEOUT
import java.time.Duration

interface IntegrationTest {

    companion object {

        val container = PostgreSQLContainer("postgres:latest")
            .waitingFor(
                WaitAllStrategy(WITH_MAXIMUM_OUTER_TIMEOUT)
                    .withStartupTimeout(Duration.ofSeconds(1))
                    .withStrategy(Wait.forLogMessage(".*database system is ready to accept connections.*\\s", 2))
                    .withStrategy(Wait.forListeningPort())
            )

        @JvmStatic
        @DynamicPropertySource
        fun properties(registry: DynamicPropertyRegistry) = registry.apply {
            add("spring.datasource.url", container::getJdbcUrl)
            add("spring.datasource.username", container::getUsername)
            add("spring.datasource.password", container::getPassword)
            add("spring.datasource.driver-class-name", container::getDriverClassName)
        }

        @JvmStatic
        @BeforeAll
        fun init() = container.start()

    }

}